### ISSUE TITLE
How to train open_clip to generate embeddings for new image-text pairs?

### ISSUE THREAD

**Issue by smith-co** [mlfoundations/open_clip#114]
How to fine tune this model to generate embeddings for new image-text pairs? Can anyone provide me with code snippets or examples?

**Comment by carlini** [comment]
I am confused by your question. What do you mean "fine tune this model to generate embeddings for new image-text pairs"?

**Comment by smith-co** [comment]
I want to use `open_clip` on a very `specific` dataset to detect near duplicate image. However, `open_clip` has not seen those images before. So I am thinking to apply finetune the `open_clip` model in my specific dataset.

**Comment by carlini** [comment]
1. The benefit of clip is that it works remarkably well on images it's never seen before. I would try to just perform standard classification before you fine-tune to see if it's even necessary.

2. If you actually do want to fine-tune the model, you might want to look at the Readme section titled "Training CLIP" and "Fine-tuning on classification tasks". Please feel free to open an issue if you have a specific problem with following these instructions.

**Comment by smith-co** [comment]
@carlini for my case clip is not giving good result for duplicate image detection as it is application specific.

I do not want to train from scratch, instead would want to do fine-tune the mode. 

The readme states "Fine-tuning on classification tasks". However, my use case is to detect near duplicate images. So I don't really follow what should I be doing in this case.

Would really appreciate your feedback 🙏

**Comment by carlini** [comment]
This repository is intended to do CLIP-style training. The best way I know of to do duplicate detection with CLIP is then is to compute image embeddings and check if they are nearby (or basically identical). But for this you'll need image-caption pairs to train the CLIP model. So if that's not working for you, then I don't think this repository will have your answer, and it sounds more like a research question for how to make a better duplicate detection algorithm.

**Comment by smith-co** [comment]
@carlini computing `embedding` for two images and then using `cosine_similarity` worked out remarkably well for me on general images.

However, for the specific image dataset I am trying the above approach is not working out so well. I am trying to apply specific duplicate detection on a dataset where the definition of duplicate does not follow the general definition of duplicate. That's why I was thinking whether it is possible to train the model for a specific context.

I already have a a dataset of pairs of images and label whether they are duplicates or not.

img1  | img2 | is_duplicate|
-----|---- -- | -----------|
1.jpg|2.jpg | 1|
1.jpg|3.jpg |0|
3.jpg|4.jpg | 1
5.jpg | 6.jpg | 0|

**Comment by carlini** [comment]
Yeah so open_clip won't help you here to perform classification. This repository is designed to perform CLIP-style training, when you move to a different task you will need to do new fine tuning. Starting with a CLIP-pretrained model would be a good first start, but you'll have to find something else to do classification.

**Comment by smith-co** [comment]
Thanks for your input @carlini. Really appreciate it 🙏 

Trying to understand your feedback. If I interpret it correctly, you are suggesting I should look into fine tuning the CLIP-pre-trained model for my task. For fine-tuning you provided me this link https://github.com/mlfoundations/open_clip#fine-tuning-on-classification-tasks.

But what I don't understand how can I find-tune CLIP in my task/dataset. 

img1  | img2 | is_duplicate|
-----|------ | -----------|
1.jpg|2.jpg | 1|
1.jpg|3.jpg |0|
1.jpg|4.jpg |0|
1.jpg|5.jpg |0|
3.jpg|4.jpg | 1
5.jpg | 6.jpg | 0|

**Comment by mitchellnw** [comment]
The link provided for fine-tuning is for when you have a lablel `y` associated for each image `x` which is not your case. Unfortunately, I think we are not the right repository to use for your use-case. If I were you I would probably fine-tune an MLP on top of the concatenated representations for both images, but we do not have code in our repository for this.

**Comment by smith-co** [comment]
@mitchellnw thanks for your response. What MLP are you referring to here? Are you suggesting to fine tuning [vision transformers](https://huggingface.co/docs/transformers/model_doc/vit)? But isn't it also for image classification primarily?

**Comment by mitchellnw** [comment]
No I mean MLP in this sense: https://en.wikipedia.org/wiki/Multilayer_perceptron

If I were approaching this problem I would probably use pseudocode like following (I emphasize that it is pseudocode and wont actually run but hopefully gives an idea)
```
MLP = torch.nn.Sequential(nn.Linear(d_feature, d_hidden), nn.ReLU(), nn.Linear(d_hidden, d_out))
backbone_model = ... # some good off the shelf pre-trained model

for img1, img2, label:
    img1_feature = backbone_model(img1).detach()
    img2_feature = backbone_model(img2).detach()
    predicted_label = MLP(torch.cat(img1_feature, img2_feature))
    loss = loss_fn(predicted_label, label)
    loss.backward()
```

**Comment by mitchellnw** [comment]
Unfortunately this is not the purpose of this repository, best of luck!

**Comment by rom1504** [comment]
If you have a large set of matched duplicate images (millions), you could use a clip-like architecture : use the same contrastive loss but instead of having (image, text) encoders you can have (image, image) ones.

That's only one option though, there is a broad literature on the topic, including Siamese nets, triplet loss,...

You can decide to train a model such that similarity (image1, image2) being higher means duplicate. You can also decide to just train this as a classification problem but it might result in worse results.


In all cases, you will definitely need to change the code.

**Comment by rom1504** [comment]
If you really want to just fine tune a clip, a last option for you is to try and get an (image, text) dataset that looks like your image distribution, for example by extracting such image text samples from a large dataset using similarity with your images.
